src/casscf/casscf_gradient.cc

```diff
@@ -0,0 +1,477 @@
+ * (4), Write 1RDMs and 2RDMs coefficients;
+ * (5), Back-transform the TPDM. 
+ */
+
+#include "casscf/casscf.h"
+#include "gradient_tpdm/backtransform_tpdm.h"
+#include "ambit/tensor.h"
+#include "ambit/blocked_tensor.h"
+
+#include "integrals/integrals.h"
+#include "base_classes/rdms.h"
+
+#include "psi4/libfock/jk.h"
+#include "psi4/libmints/matrix.h"
+#include "psi4/libmints/wavefunction.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/psifiles.h"
+
+#include "helpers/printing.h"
+#include "helpers/helpers.h"
+#include "helpers/timer.h"
+#include "sci/aci.h"
+#include "fci/fci_solver.h"
+#include "base_classes/active_space_solver.h"
+
+#include "sci/fci_mo.h"
+#include "orbital-helpers/mp2_nos.h"
+#include "orbital-helpers/orbitaloptimizer.h"
+#include "orbital-helpers/semi_canonicalize.h"
+
+#ifdef HAVE_CHEMPS2
+#include "dmrg/dmrgsolver.h"
+#endif
+#include "psi4/libdiis/diisentry.h"
+#include "psi4/libdiis/diismanager.h"
+#include "psi4/libmints/factory.h"
+#include "psi4/psifiles.h"
+#include "psi4/libiwl/iwl.hpp"
+#include "psi4/libpsio/psio.hpp"
+
+using namespace ambit;
+using namespace psi;
+
+namespace forte 
+{
+
+
+/**
+ * Initialize global variables. 
+ */
+void CASSCF::set_all_variables()
+{
+    // Set MOs containers.
+    core_mos_ = mo_space_info_->get_corr_abs_mo("RESTRICTED_DOCC");
+    actv_mos_ = mo_space_info_->get_corr_abs_mo("ACTIVE");
+    virt_mos_ = mo_space_info_->get_corr_abs_mo("RESTRICTED_UOCC");
+    core_all_ = mo_space_info_->get_absolute_mo("RESTRICTED_DOCC");
+    actv_all_ = mo_space_info_->get_absolute_mo("ACTIVE");
+    core_mos_relative = mo_space_info_->get_relative_mo("RESTRICTED_DOCC");
+    actv_mos_relative = mo_space_info_->get_relative_mo("ACTIVE");
+    irrep_vec = mo_space_info_->get_dimension("ALL");
+    
+    // Set MO spaces.
+    set_ambit_space();
+    
+    // Initialize tensors.
+    Gamma1_ = BTF_->build(CoreTensor, "Gamma1", spin_cases({"aa"}));
+    Gamma2_ = BTF_->build(CoreTensor, "Gamma2", spin_cases({"aaaa"}));
+    H_ = BTF_->build(CoreTensor, "One-Electron Integral", spin_cases({"gg"}));
+    V_ = BTF_->build(CoreTensor, "Electron Repulsion Integral", spin_cases({"gggg"}));
+    F_ = BTF_->build(CoreTensor, "Fock Matrix", spin_cases({"gg"}));
+    W_ = BTF_->build(CoreTensor, "Lagrangian", spin_cases({"gg"}));
+
+    // Set tensors.
+    set_tensor();
+}
+
+
+/**
+ * Set ambit MO spaces. 
+ */
+void CASSCF::set_ambit_space() 
+{
+    outfile->Printf("\n    Setting Ambit MO Space .......................... ");
+    
+    BlockedTensor::reset_mo_spaces();
+    BlockedTensor::set_expert_mode(true);
+    BTF_= std::make_shared<BlockedTensorFactory>();
+    std::string acore_label_ = "c";
+    std::string aactv_label_ = "a";
+    std::string avirt_label_ = "v";
+    std::string bcore_label_ = "C";
+    std::string bactv_label_ = "A";
+    std::string bvirt_label_ = "V";
+
+    // Add Ambit index labels.
+    BTF_->add_mo_space(acore_label_, "mn$%", core_mos_, AlphaSpin);
```

Comment:
Can you change core labels to m, n, c0, c1, c2, ... instead of characters like !@# and make changes to other labels as well. I know you did not start this but I do think they look hideous and hard to interpret.

Reply:
I have fixed this.